instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/SpringDeclarativeConfigProperties.java

```diff
@@ -0,0 +1,316 @@
+  private final ComponentLoader componentLoader;
+
+  private SpringDeclarativeConfigProperties(
+      Map<String, Object> simpleEntries,
+      Map<String, List<SpringDeclarativeConfigProperties>> listEntries,
+      Map<String, SpringDeclarativeConfigProperties> mapEntries,
+      ComponentLoader componentLoader) {
+    this.simpleEntries = simpleEntries;
+    this.listEntries = listEntries;
+    this.mapEntries = mapEntries;
+    this.componentLoader = componentLoader;
+  }
+
+  /**
+   * Create a {@link SpringDeclarativeConfigProperties} from the {@code properties} map.
+   *
+   * <p>{@code properties} is expected to be the output of YAML parsing (i.e. with Jackson {@code
+   * com.fasterxml.jackson.databind.ObjectMapper}), and have values which are scalars, lists of
+   * scalars, lists of maps, and maps.
+   *
+   * @see DeclarativeConfiguration#toConfigProperties(Object)
+   */
+  @SuppressWarnings("unchecked")
+  public static SpringDeclarativeConfigProperties create(
+      Map<String, Object> properties, ComponentLoader componentLoader) {
+    Map<String, Object> simpleEntries = new LinkedHashMap<>();
+    Map<String, List<SpringDeclarativeConfigProperties>> listEntries = new LinkedHashMap<>();
+    Map<String, SpringDeclarativeConfigProperties> mapEntries = new LinkedHashMap<>();
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      String key = entry.getKey();
+      Object value = entry.getValue();
+      if (isPrimitive(value) || value == null) {
+        simpleEntries.put(key, value);
+        continue;
+      }
+      if (isPrimitiveList(value)) {
+        simpleEntries.put(key, value);
+        continue;
+      }
+      if (isListOfMaps(value)) {
+        List<SpringDeclarativeConfigProperties> list =
+            ((List<Map<String, Object>>) value)
+                .stream()
+                    .map(map -> SpringDeclarativeConfigProperties.create(map, componentLoader))
+                    .collect(toList());
+        listEntries.put(key, list);
+        continue;
+      }
+      if (isMap(value)) {
+        SpringDeclarativeConfigProperties configProperties =
+            SpringDeclarativeConfigProperties.create((Map<String, Object>) value, componentLoader);
+        mapEntries.put(key, configProperties);
+        continue;
+      }
+      throw new DeclarativeConfigException(
+          "Unable to initialize ExtendedConfigProperties. Key \""
+              + key
+              + "\" has unrecognized object type "
+              + value.getClass().getName());
+    }
+    return new SpringDeclarativeConfigProperties(
+        simpleEntries, listEntries, mapEntries, componentLoader);
+  }
+
+  private static boolean isPrimitiveList(Object object) {
+    if (object instanceof List) {
+      List<?> list = (List<?>) object;
+      return list.stream().allMatch(SpringDeclarativeConfigProperties::isPrimitive);
+    }
+    return false;
+  }
+
+  private static boolean isPrimitive(Object object) {
```

Comment:
since `String` is not primitive the name is slightly misleading. Perhaps something like `isSimple` would be a better fit.
Also it does not list all the primitive types.